### PR TITLE
feat: implement vote manager logics

### PR DESCRIPTION
**Pull request by DNK90**

### CODE REVIEW

File: core/vote.go
Comment (minh-bq) [inline core/vote.go:1]: This file is the same file as core/types/vote.go above

File: core/types/vote.go
Comment (minh-bq) [inline core/types/vote.go:1]: Can you make this file into a separate PR so we can merge this file before?

File: core/types/vote.go [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+	"github.com/ethereum/go-ethereum/crypto/bls"
+	"github.com/pkg/errors"
+
+	"github.com/ethereum/go-ethereum/common"
+)
+
+const (
+	BLSPublicKeyLength = 48
+	BLSSignatureLength = 96
```
Comment (minh-bq) [inline core/types/vote.go:14]: These constants are in params/config.go already, so we don't need to redefine them here

File: core/types/vote.go [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+	MaxAttestationExtraLength = 256
```
Comment (minh-bq) [inline core/types/vote.go:16]: Is this variable unused?

File: core/types/vote.go [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+	hash atomic.Value
+}
+
+// VoteAttestation represents the votes of super majority validators.
+type VoteAttestation struct {
+	VoteAddressSet ValidatorsBitSet // The bitset marks the voted validators.
+	AggSignature   BLSSignature     // The aggregated BLS signature of the voted validators' signatures.
+	Data           *VoteData        // The vote data for fast finality.
+	Extra          []byte           // Reserved for future usage.
+}
```
Comment (minh-bq) [inline core/types/vote.go:48]: Is this struct unused?

File: core/types/vote.go [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+const (
+	BLSPublicKeyLength = 48
+	BLSSignatureLength = 96
+
+	MaxAttestationExtraLength = 256
+)
+
+type BLSPublicKey [BLSPublicKeyLength]byte
+type BLSSignature [BLSSignatureLength]byte
```
Comment (minh-bq) [inline core/types/vote.go:20]: These types are in crypto/bls package already, we shouldn't redefine them here

File: core/types/vote.go [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+	TargetNumber uint64      // The target block number which validator wants to vote for.
+	TargetHash   common.Hash // The block hash of the target block.
+}
+
+// Hash returns the hash of the vote data.
+func (d *VoteData) Hash() common.Hash { return rlpHash(d) }
+
+// VoteEnvelope represents the vote of a single validator.
+type VoteEnvelope struct {
+	VoteAddress BLSPublicKey // The BLS public key of the validator.
```
Comment (minh-bq) [inline core/types/vote.go:34]: I think this should be renamed to PublicKey, address is not a terminology in BLS signature scheme.

File: core/types/vote.go (hunk 4 above)
Comment (DNK90) [reply core/types/vote.go:20]: these are purely slices not the object itself

File: core/types/vote.go (hunk 5 above)
Comment (DNK90) [reply core/types/vote.go:34]: I think the vote is using bls public key, therefore, remove the `BLS` would make confusion

Comment (minh-bq) [reply core/types/vote.go:34]: I mean we should change the name of the variable from VoteAddress to PublicKey

File: core/types/vote.go (hunk 4 above)
Comment (minh-bq) [reply core/types/vote.go:20]: But why don't we use common.PublicKey and common.Signature from crypto/bls/common/interface.go?

Comment (DNK90) [reply core/types/vote.go:20]: we also store pubKey and secret key at slice of bytes in keymanager. Therefore we keep these type at byte also

File: core/vote/vote_pool.go [hunk 6]
```diff
@@ -0,0 +1,397 @@
...
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/event"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/metrics"
+	"github.com/ethereum/go-ethereum/params"
+)
+
+const (
+	maxCurVoteAmountPerBlock    = 21
```
Comment (minh-bq) [inline core/vote/vote_pool.go:20]: This must be greater or equal to 22, not sure if we want more than 22 validators in the future.

File: core/vote/vote_pool_test.go [hunk 7]
```diff
@@ -0,0 +1,430 @@
...
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
```
Comment (minh-bq) [inline core/vote/vote_pool_test.go:15]: Not sure about the license part here, we may need to remove this

File: core/vote/vote_pool.go (hunk 6 above)
Comment (DNK90) [reply core/vote/vote_pool.go:20]: we should move this to config

File: core/vote/vote_journal.go [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+	"github.com/tidwall/wal"
+
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/metrics"
+)
+
+const (
+	maxSizeOfRecentEntry    = 512
+	maliciousVoteSlashScope = 256
```
Comment (minh-bq) [inline core/vote/vote_journal.go:16]: Look like unused constant
